### ISSUE TITLE
Custom error response

### ISSUE THREAD

**Issue by jacob-pro** [actix/actix-extras#13]
Is there any way to catch and transform the authentication errors that are generated by the middleware before they get sent as a response?

If not do you think this would be best implemented by adding a error_handler to the middleware initialisation, or maybe having an alternative signature for the validator function that takes a `cred: Result<BasicAuth, Error>` ?

**Comment by tespent** [comment]
I recently encounter exactly the same issue, hoping to change the response body of `AuthenticationError` but failed. 

Is it possible to make AuthenticationError a trait, or simply add some callbacks like `complete_response` to replace `.finish()` here?

https://github.com/actix/actix-extras/blob/81c48930c4c89a2856bfc37ba6feb02040a93d36/actix-web-httpauth/src/extractors/errors.rs#L58

Conceptual usage:

```rust
// struct AuthenticationError<B: CompleteErrorResponse>

#[derive(Serialize)]
enum ApiStatus{
  Unauthoirzed, Ok
}

#[derive(Debug, Clone, Default)]
struct ApiErrorResponse {}
impl CompleteErrorResponse for ApiErrorResponse {
  fn complete_response(builder: &mut HttpResponseBuilder) -> HttpResponse {
    builder.json(ApiStatus::Unauthorized)
  }
}

type MyBasicAuth = BasicAuth<ApiErrorResponse>;

#[get("/api/")]
fn api_index(credential: MyBasicAuth) -> impl Responder {
  debug!("Hello, {}", credential.user_id());
  Response::Ok().json(ApiStatus::Ok)
}
```

**Comment by robjtede** [comment]
should be doable now with #205; released in actix-web-httpauth v0.6.0-beta.4

**Comment by jacob-pro** [comment]
Thanks for adding this, I will start using as soon as my other dependencies catch up to actix 4.

As far as I understand I just have to change my usage from:

```
pub async fn authenticator(req: ServiceRequest, auth: BearerAuth) -> Result<ServiceRequest, actix_web::Error>;
let auth = HttpAuthentication::bearer(authenticator);
```

to:


```
pub async fn authenticator(req: ServiceRequest, auth: Option<BearerAuth>) -> Result<ServiceRequest, actix_web::Error>;
let auth = HttpAuthentication::with_fn(authenticator);
```
